### DISCUSSION TITLE
Replacing the WebExtensions portal

### DISCUSSION THREAD

**Discussion by swick** [flatpak/xdg-desktop-portal#1191] [Ideas]
With improvements in dbus it should become possible it the future again for flatpaked or otherwise sandboxed apps to talk to each other. When support for this lands, we can add a new native messaging system to browsers which uses dbus to open the pipe, as outlined in this comment: https://github.com/flatpak/xdg-desktop-portal/issues/705#issuecomment-1794890805

The need for a WebExtensions portal will then be limited to apps native messaging servers which did not upgrade to the new mechanism and can eventually be removed.

**Comment by fenglengshun** [comment]
Would this help with Discord IPC and Currently Playing as well? I tried the Discord Flatpak but couldn't find the game I want to show on my Activity through their process selector. The alternative was using arrpc and vencord but that still required poking some holes in the sandbox too.

**Reply by TingPing** [reply]
Not related to Currently Playing. That needs a host service that does that work for Discord, be it a portal or otherwise.

**Comment by Mikenux** [comment]
Why not? But:
- It is a Web Browser that wants to communicate with a native app, not an extension, the WB being not reliable from the point of view of a sandboxing system. Access approval is therefore required (regardless of what needs to be said to the user).
- Apps without network access will have access to the network thanks to this talk permission. A portal UI is therefore necessary to inform the user about the possibility of sharing private information (because we cannot control the type of information that the WB and the app will exchange).
- The user must be informed if the WB wants to launch an app, because an extension could possibly be installed automatically by a vendor/distributor/administrator.
- We may want to revoke the talk permission between a WB and an app when a user removes the extension from the WB. I think it makes sense. How can we handle this, since the WB can say that an extension has been removed, even if it hasn't?

**Reply by brunoais** [reply]
I think there's no need for that much. I think it should just allow the browser to manage this process and trust it to manage once the user says it's OK for the browser to do that management.
If the user wants to do a more fine tune of it, then allow the user to do it but the management of what parts of inside of the sandbox can communicate outside of as native messaging should be managed by the browser. The user already has to give consent in the browser when installing the extension.

**Reply by Mikenux** [reply]
Like said, the web browser is not reliable / trusted from the point of view of a sandboxing system (e.g. flatpak, snap). Agreeing once for all the management means allowing the web browser to communicate with apps at will (however, the user may detect unwanted behavior if we show a notification if an app is launched in the background). Note that the user can "always allow" when asked.

> The user already has to give consent in the browser when installing the extension.

To grant access by installing an extension, we need to be in control of this process, as, again, the web browser is not trusted.

**Reply by swick** [reply]
Apps can already communicate with each other over a number of channels (udp, tcp, tmp, ...). None of that is security relevant. If an app wants to share information with another app, it can do so. The only difference with dbus is going to be that the both parties are authenticated, so instead of sharing it with everyone, one can have a policy in an app. A password manager could say "I trust firefox and will give firefox access to all passwords" or "I don't know this app, I will ask the user if I should give it access to this password".

**Reply by brunoais** [reply]
My idea is more inclined to having a selection of directories/files which can be executed from inside the sandbox the same way there's the readonly/readwrite option for the sandbox. This would be an execute permission instead.

**Reply by Mikenux** [reply]
@swick: Even though there are no security issues, there is privacy. Any app intended to be fully sandboxed cannot communicate with any other. But, regardless of an app's sandbox state, the only parties that can grant permission are the user and the system (if the system has a policy for "privileged" apps, decided by the distributor or an administrator, which seems not to be implemented at the moment). An app cannot decide.

@brunoais: What are you talking about?

**Reply by swick** [reply]
> Any app intended to be fully sandboxed cannot communicate with any other.

That's just wrong.

**Reply by Mikenux** [reply]
If a fully sandboxed app can communicate with other apps, then there is a problem.

**Reply by swick** [reply]
There is not.

**Reply by Mikenux** [reply]
Another aspect to consider is that the permission is supposed to be revoked when an extension is disabled.

**Comment by iavael** [comment]
Why would authors of hundreds of webextensions care to support linux-specific non-portable messaging system?
What's the plan to motivate extension developers to do extra work supporting dbus-based messaging?
Because even if you make major browsers support dbus-based messaging (which is quite a work by itself), low adoption among extension developers would make this initiative dead on arrival.

**Reply by swick** [reply]
1. The changes are transparent for the extension.
2. Browsers already have to implement this mechanism for every OS, because it is OS specific
3. Starting apps by executing a file at a specific path is simply not the proper way to start apps on the linux desktop and has not been for decades.
4. D-Bus activation solves the problem of apps being confined and or in containers where a container specific path translation is otherwise necessary
5. D-Bus activation solves the problem of putting the processes in the right cgroup and systemd unit

**Reply by BurstingBrain** [reply]
Well if it is ever so simple to do that, why don't you patch Chromium and Firefox, make it work and then we'll all live in a more secure place thank to you. Meanwhile just let me have my damn password manager integration, my DE integration, and my external downloader. Because even though I love Debian. I want *some* fresh apps, that work together without having to bash my head against the wall. I already have to wait for the next Debian release, quit arguing meaninglessly, and do some code instead. By looking to far into the future you forget the present 🫥

**Reply by filips123** [reply]
I agree with @iavael. Currently, the actual communication is done through stdin/stdout, which works basically the same in all operating systems, and is supported in virtually every programming language, without any extra code or libraries. Handling D-Bus is more complicating and often requires additional libraries, if any even exist for the language the native messaing cost is written in.

> The changes are transparent for the extension.

They aren't transparent to the native messaging host, which is commonly developed by the same people as the extension. So, it's not really transparent to them.

**Comment by Mikenux** [comment]
Wouldn't it be better to have an app that indicates which app to communicate with than to have to indicate an extension? While I don't know how Discord operates, it will allow apps to ask to communicate with Discord, because I doubt Discord is using a list of allowed apps. This will allow this portal to do more things, and avoid developing something specifically for web browsers, namely searching for the app that has the extension given by the browser as allowed. However, this would mean that extensions would have to indicate which app to communicate with. As an advantage, this allows developers to have a trusted model where a specific app can be specified as the app ID + repository + branch, for example.

**Reply by Mikenux** [reply]
@orowith2os

Seen [your comment](https://github.com/flatpak/xdg-desktop-portal/issues/653#issuecomment-1364634800). Is Discord limited to Rich Presence or is there more than that?

**Reply by fenglengshun** [reply]
@Mikenux I think so, but there's a lot to Rich Presence - like [arrpc](https://github.com/OpenAsar/arrpc), [PreMiD](https://premid.app), a different [Rich Presence extension](https://github.com/lolamtisch/Discord-RPC-Extension), the [IPC Bridge](https://github.com/sonic2kk/steamtinkerlaunch/wiki/Discord-IPC-Bridge-(WDIB)), [mpv-discordRPC](https://github.com/cniw/mpv-discordRPC).

**Reply by Mikenux** [reply]
Thanks. This seems only about text data for statuses. If there was some command exchange other than for a status, then Discord is a case. Otherwise, not. If it is only for that, having a portal where data can be parsed and limiting amount of data that can be exchanged is better (or with use of metadata, maybe with wayland), as here we assume any data can be exchanged between apps. For web browsers, limits are max 1 MB from app to browser - sufficiently big for some file transmission - and max 4 GB from browser to app... (and I wonder what are the cases for that...).

**Reply by Mikenux** [reply]
@Rob--W: A thought on what I describe in the first comment on this thread? In addition, could you tell which cases should be able to send up to 4 GB of data through a web browser? (this is to find out if it is possible to have limits in the portal and which ones)

**Reply by rugk** [reply]
[Backlink to a user having the same problem with LibreOffice + integration in Zotero/Firefox](https://github.com/flatpak/xdg-desktop-portal/issues/655#issuecomment-2683527420).

Anyway, my personal opinion is: Let's get that PR for NativeMessaging merged first, finally and not get distracted in these discussions. You know, perfect is the enemy of good and we're waiting for this very long, already. :wink:

**Reply by Mikenux** [reply]
@rugk: I know :wink: But, here, it is for the long-term solution (I forgot to say it), so unrelated to the PR which is for the short-term solution ("quick" fix). The discussion can therefore continue.

**Reply by Mikenux** [reply]
App Request: extension vs. app + extension:
- extension. The portal will search for an app having the extension. Any app having this extension in its allow list will be shown.
- app + extension. The portal will directly search for the app and check if it has the extension in its allow list. This has the advantage that the request can be denied if the app does not exist or when the app does not have the extension in its list (if it exists).

**Comment by Mikenux** [comment]
Another thing: Data exchange allowed when app has the focus. What would be interesting to know is if it is only for the requesting app or also for the requested app.

**Reply by Mikenux** [reply]
For that, the app needs to be aware when it has the focus.

Additionally, trust checking between apps must take place every time the request (for communication/data exchange) is made, so that trust revocation by any of the apps can be handled.

**Comment by Mikenux** [comment]
I think what is also wanted is to be able to disallow the use of portal permissions and/or to redisplay portal windows when an app receives data from another app.

**Comment by Mikenux** [comment]
There is a problem I just thought about.

I designed my proposals for neighboring files so that an app uses the right proposal for its case: it's case exclusivity. Here, native messaging allows to get credentials and, I guess, to exchange files. There were already discussion and code for a credential portal, and we know that a portal should be used to exchange files between apps. Thus, with native messaging, it does not seem possible to ensure case exclusivity. I don't know, at least currently, how this could be handled (I'm just reporting the problem).
